pkg/remote/builtin/gitlab/gitlab.go

```diff
@@ -259,20 +230,66 @@ func (r *Gitlab) Hook(req *http.Request) (*common.Hook, error) {
 		return nil, err
 	}
 
-	if len(parsed.After) == 0 || parsed.TotalCommitsCount == 0 {
-		return nil, nil
-	}
+	switch parsed.ObjectKind {
+	case "merge_request":
+		if parsed.ObjectAttributes.State != "reopened" && parsed.ObjectAttributes.MergeStatus != "unchecked" ||
+			parsed.ObjectAttributes.State != "opened" && parsed.ObjectAttributes.MergeStatus != "unchecked" {
+			return nil, nil
+		}
+
+		return mergeRequest(parsed, req)
+	case "tag_push", "push":
+		if len(parsed.After) == 0 || parsed.TotalCommitsCount == 0 {
+			return nil, nil
+		}
 
-	if parsed.ObjectKind == "merge_request" {
-		// NOTE: in gitlab 8.0, gitlab will get same MR models as github
-		//       https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/981/diffs
+		return push(parsed, req)
+	default:
 		return nil, nil
 	}
+}
 
-	if len(parsed.After) == 0 {
-		return nil, nil
+func mergeRequest(parsed *gogitlab.HookPayload, req *http.Request) (*common.Hook, error) {
+	var hook = new(common.Hook)
+
+	re := regexp.MustCompile(".git$")
+
+	hook.Repo = &common.Repo{}
+	hook.Repo.Owner = req.FormValue("owner")
+	hook.Repo.Name = req.FormValue("name")
+	hook.Repo.FullName = fmt.Sprintf("%s/%s", hook.Repo.Owner, hook.Repo.Name)
+	hook.Repo.Link = re.ReplaceAllString(parsed.ObjectAttributes.Target.HttpUrl, "$1")
```

Comment:
Yeah, I was suggesting that you can add it to the GitLab `hook_attrs` method on `Project`, so that you can use it here.